How to parallelize?

I use `javaView.getClasses().parallelStream()` to retrieve some information for all classes.

But it throws `java.util.ConcurrentModificationException`:
```
Caused by: java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1229)
	at sootup.java.core.JavaIdentifierFactory.getPackageName(JavaIdentifierFactory.java:254)
Caused by: java.util.ConcurrentModificationException
	at sootup.java.core.JavaIdentifierFactory.getClassType(JavaIdentifierFactory.java:108)
	at sootup.java.core.JavaIdentifierFactory.getClassType(JavaIdentifierFactory.java:122)
	at sootup.java.core.JavaIdentifierFactory.lambda$getType$1(JavaIdentifierFactory.java:175)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at sootup.java.core.JavaIdentifierFactory.getType(JavaIdentifierFactory.java:175)
	at sootup.java.bytecode.frontend.AsmUtil.toJimpleType(AsmUtil.java:200)
	at sootup.java.bytecode.frontend.AsmClassSource.lambda$resolveFields$0(AsmClassSource.java:63)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at sootup.java.bytecode.frontend.AsmClassSource.resolveFields(AsmClassSource.java:75)
	at sootup.java.bytecode.frontend.AsmClassSource.resolveFields(AsmClassSource.java:147)
	at sootup.core.model.SootClass.lazyFieldInitializer(SootClass.java:63)
	at com.google.common.base.Suppliers$NonSerializableMemoizingSupplier.get(Suppliers.java:166)
	at sootup.core.model.SootClass.getFields(SootClass.java:106)
	at sootup.java.core.JavaSootClass.getFields(JavaSootClass.java:109)
```

It seems that `sootup.java.core.JavaIdentifierFactory` is still a global singleton and in this class `java.util.HashMap` is used for cache, which doesn't support concurrency.

Is that exception expected? What is the correct way to parallelize?

--- a/sootup.java.core/src/main/java/sootup/java/core/JavaIdentifierFactory.java
+++ b/sootup.java.core/src/main/java/sootup/java/core/JavaIdentifierFactory.java
@@ -59,10 +59,14 @@ public class JavaIdentifierFactory implements IdentifierFactory {
   @Nonnull private static final JavaIdentifierFactory INSTANCE = new JavaIdentifierFactory();
 
   /** Caches the created PackageNames for packages. */
-  @Nonnull protected final Cache<String, PackageName> packages = CacheBuilder.newBuilder().weakValues().build();
+  @Nonnull
+  protected final Cache<String, PackageName> packages =
+      CacheBuilder.newBuilder().weakValues().build();
 
   /** Caches annotation types */
-  @Nonnull protected final Cache<String, AnnotationType> annotationTypes = CacheBuilder.newBuilder().weakValues().build();
+  @Nonnull
+  protected final Cache<String, AnnotationType> annotationTypes =
+      CacheBuilder.newBuilder().weakValues().build();
 
   @Nonnull
   protected final Map<String, PrimitiveType> primitiveTypeMap = Maps.newHashMapWithExpectedSize(8);
@@ -212,8 +216,11 @@ public class JavaIdentifierFactory implements IdentifierFactory {
     String className = ClassUtils.getShortClassName(fullyQualifiedClassName);
     String packageName = ClassUtils.getPackageName(fullyQualifiedClassName);
 
-    return annotationTypes.asMap().computeIfAbsent(
-        className + packageName, (k) -> new AnnotationType(className, getPackageName(packageName)));
+    return annotationTypes
+        .asMap()
+        .computeIfAbsent(
+            className + packageName,
+            (k) -> new AnnotationType(className, getPackageName(packageName)));
   }
 
   @Override

--- a/sootup.java.core/src/main/java/sootup/java/core/JavaIdentifierFactory.java
+++ b/sootup.java.core/src/main/java/sootup/java/core/JavaIdentifierFactory.java
@@ -22,6 +22,8 @@ package sootup.java.core;
  * #L%
  */
 
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
 import com.google.common.collect.Maps;
 import java.nio.file.Path;
 import java.util.*;
@@ -57,10 +59,10 @@ public class JavaIdentifierFactory implements IdentifierFactory {
   @Nonnull private static final JavaIdentifierFactory INSTANCE = new JavaIdentifierFactory();
 
   /** Caches the created PackageNames for packages. */
-  @Nonnull protected final Map<String, PackageName> packages = new HashMap<>();
+  @Nonnull protected final Cache<String, PackageName> packages = CacheBuilder.newBuilder().weakValues().build();
 
   /** Caches annotation types */
-  @Nonnull protected final Map<String, AnnotationType> annotationTypes = new HashMap<>();
+  @Nonnull protected final Cache<String, AnnotationType> annotationTypes = CacheBuilder.newBuilder().weakValues().build();
 
   @Nonnull
   protected final Map<String, PrimitiveType> primitiveTypeMap = Maps.newHashMapWithExpectedSize(8);
@@ -210,7 +212,7 @@ public class JavaIdentifierFactory implements IdentifierFactory {
     String className = ClassUtils.getShortClassName(fullyQualifiedClassName);
     String packageName = ClassUtils.getPackageName(fullyQualifiedClassName);
 
-    return annotationTypes.computeIfAbsent(
+    return annotationTypes.asMap().computeIfAbsent(
         className + packageName, (k) -> new AnnotationType(className, getPackageName(packageName)));
   }
 
@@ -251,7 +253,7 @@ public class JavaIdentifierFactory implements IdentifierFactory {
    */
   @Override
   public PackageName getPackageName(@Nonnull final String packageName) {
-    return packages.computeIfAbsent(packageName, (name) -> new PackageName(name));
+    return packages.asMap().computeIfAbsent(packageName, PackageName::new);
   }
 
   /**

--- a/sootup.java.core/src/main/java/sootup/java/core/JavaModuleIdentifierFactory.java
+++ b/sootup.java.core/src/main/java/sootup/java/core/JavaModuleIdentifierFactory.java
@@ -36,7 +36,8 @@ public class JavaModuleIdentifierFactory extends JavaIdentifierFactory {
 
   public static final String MODULE_INFO_FILE = "module-info";
 
-  private static final Cache<String, ModuleSignature> modules = CacheBuilder.newBuilder().weakValues().build();
+  private static final Cache<String, ModuleSignature> modules =
+      CacheBuilder.newBuilder().weakValues().build();
 
   private static final JavaModuleIdentifierFactory INSTANCE = new JavaModuleIdentifierFactory();
 
@@ -52,8 +53,9 @@ public class JavaModuleIdentifierFactory extends JavaIdentifierFactory {
       moduleIdentifierFactoryWrapper = CacheBuilder.newBuilder().weakValues().build();
 
   public static JavaModuleIdentifierFactory getInstance(@Nonnull ModuleSignature moduleSignature) {
-    return moduleIdentifierFactoryWrapper.asMap().computeIfAbsent(
-            moduleSignature, JavaModuleIdentifierFactoryWrapper::new);
+    return moduleIdentifierFactoryWrapper
+        .asMap()
+        .computeIfAbsent(moduleSignature, JavaModuleIdentifierFactoryWrapper::new);
   }
 
   static {
@@ -160,15 +162,20 @@ public class JavaModuleIdentifierFactory extends JavaIdentifierFactory {
   public ModulePackageName getPackageName(
       @Nonnull final String packageName, @Nonnull final String moduleName) {
     String fqId = moduleName + "." + packageName;
-    return (ModulePackageName) packages.asMap().computeIfAbsent(
-            fqId, key -> new ModulePackageName(packageName, getModuleSignature(moduleName)));
+    return (ModulePackageName)
+        packages
+            .asMap()
+            .computeIfAbsent(
+                fqId, key -> new ModulePackageName(packageName, getModuleSignature(moduleName)));
   }
 
   public ModulePackageName getPackageName(
       @Nonnull final String packageName, @Nonnull final ModuleSignature moduleSignature) {
     String fqId = moduleSignature.getModuleName() + "." + packageName;
-    return (ModulePackageName) packages.asMap().computeIfAbsent(
-            fqId, key -> new ModulePackageName(packageName, moduleSignature));
+    return (ModulePackageName)
+        packages
+            .asMap()
+            .computeIfAbsent(fqId, key -> new ModulePackageName(packageName, moduleSignature));
   }
 
   /** Wrapper which refers to a given ModuleSignature when building stuff */

--- a/sootup.java.core/src/main/java/sootup/java/core/JavaModuleIdentifierFactory.java
+++ b/sootup.java.core/src/main/java/sootup/java/core/JavaModuleIdentifierFactory.java
@@ -22,9 +22,9 @@ package sootup.java.core;
  * #L%
  */
 
-import java.util.HashMap;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
 import java.util.List;
-import java.util.Map;
 import javax.annotation.Nonnull;
 import org.apache.commons.lang3.ClassUtils;
 import sootup.core.signatures.MethodSignature;
@@ -36,7 +36,7 @@ public class JavaModuleIdentifierFactory extends JavaIdentifierFactory {
 
   public static final String MODULE_INFO_FILE = "module-info";
 
-  private static final Map<String, ModuleSignature> modules = new HashMap<>();
+  private static final Cache<String, ModuleSignature> modules = CacheBuilder.newBuilder().weakValues().build();
 
   private static final JavaModuleIdentifierFactory INSTANCE = new JavaModuleIdentifierFactory();
 
@@ -48,13 +48,12 @@ public class JavaModuleIdentifierFactory extends JavaIdentifierFactory {
     return getInstance(getModuleSignature(module));
   }
 
-  private static final Map<ModuleSignature, JavaModuleIdentifierFactory>
-      moduleIdentifierFactoryWrapper = new HashMap<>();
+  private static final Cache<ModuleSignature, JavaModuleIdentifierFactory>
+      moduleIdentifierFactoryWrapper = CacheBuilder.newBuilder().weakValues().build();
 
   public static JavaModuleIdentifierFactory getInstance(@Nonnull ModuleSignature moduleSignature) {
-    return moduleIdentifierFactoryWrapper.computeIfAbsent(
-        moduleSignature,
-        methodSignature -> new JavaModuleIdentifierFactoryWrapper(moduleSignature));
+    return moduleIdentifierFactoryWrapper.asMap().computeIfAbsent(
+            moduleSignature, JavaModuleIdentifierFactoryWrapper::new);
   }
 
   static {
@@ -139,12 +138,7 @@ public class JavaModuleIdentifierFactory extends JavaIdentifierFactory {
    *     the unnamed module.
    */
   public static ModuleSignature getModuleSignature(@Nonnull final String moduleName) {
-    ModuleSignature moduleSignature = modules.get(moduleName);
-    if (moduleSignature == null) {
-      moduleSignature = new ModuleSignature(moduleName);
-      modules.put(moduleName, moduleSignature);
-    }
-    return moduleSignature;
+    return modules.asMap().computeIfAbsent(moduleName, ModuleSignature::new);
   }
 
   @Override
@@ -166,24 +160,15 @@ public class JavaModuleIdentifierFactory extends JavaIdentifierFactory {
   public ModulePackageName getPackageName(
       @Nonnull final String packageName, @Nonnull final String moduleName) {
     String fqId = moduleName + "." + packageName;
-    ModulePackageName packageSignature = (ModulePackageName) packages.get(fqId);
-    if (packageSignature == null) {
-      ModuleSignature moduleSignature = getModuleSignature(moduleName);
-      packageSignature = new ModulePackageName(packageName, moduleSignature);
-      packages.put(fqId, packageSignature);
-    }
-    return packageSignature;
+    return (ModulePackageName) packages.asMap().computeIfAbsent(
+            fqId, key -> new ModulePackageName(packageName, getModuleSignature(moduleName)));
   }
 
   public ModulePackageName getPackageName(
       @Nonnull final String packageName, @Nonnull final ModuleSignature moduleSignature) {
     String fqId = moduleSignature.getModuleName() + "." + packageName;
-    ModulePackageName packageSignature = (ModulePackageName) packages.get(fqId);
-    if (packageSignature == null) {
-      packageSignature = new ModulePackageName(packageName, moduleSignature);
-      packages.put(fqId, packageSignature);
-    }
-    return packageSignature;
+    return (ModulePackageName) packages.asMap().computeIfAbsent(
+            fqId, key -> new ModulePackageName(packageName, moduleSignature));
   }
 
   /** Wrapper which refers to a given ModuleSignature when building stuff */

